magenta-lib/src/main/scala/magenta/tasks/gcp/GCSTasks.scala

```diff
@@ -0,0 +1,130 @@
+package magenta.tasks.gcp
+
+import java.io.File
+import java.math.BigInteger
+import java.net.URLConnection
+import java.util.Arrays
+
+import com.google.api.client.http.InputStreamContent
+import com.google.api.services.storage.Storage
+import com.google.api.services.storage.model.{ObjectAccessControl, StorageObject}
+import com.gu.management.Loggable
+import magenta.{DeploymentResources, KeyRing, Region, Stack, Stage}
+import magenta.artifact.{S3Location, S3Object, S3Path}
+import magenta.deployment_type.param_reads.PatternValue
+import magenta.tasks.Task
+import software.amazon.awssdk.core.internal.util.Mimetype
+import software.amazon.awssdk.services.s3.S3Client
+import software.amazon.awssdk.services.s3.model.GetObjectRequest
+
+case class GCSUpload(
+  region: Region,
+  bucket: String,
+  paths: Seq[(S3Location, String)],
+  cacheControlPatterns: List[PatternValue] = Nil,
+  extensionToMimeType: Map[String,String] = Map.empty,
+  publicReadAcl: Boolean = false,
+  detailedLoggingThreshold: Int = 10,
+)(implicit val keyRing: KeyRing, artifactClient: S3Client,
+  withClientFactory: (KeyRing, DeploymentResources) => (Storage => Unit) => Unit = GCS.withGCSClient[Unit]) extends Task with Loggable {
```

Comment:
This complicated `withClientFactory` signature exists so that a mock can be provided in tests. If you don't plan to write a test then this could be simplified.